src/Requests.php

```diff
@@ -466,11 +466,21 @@ public static function request($url, $headers = [], $data = [], $type = self::GE
 			$transport    = self::get_transport($capabilities);
 		}
 
-		$response = $transport->request($url, $headers, $data, $options);
+		try {
+			$response = $transport->request($url, $headers, $data, $options);
+
+			$options['hooks']->dispatch('requests.before_parse', [&$response, $url, $headers, $data, $type, $options]);
 
-		$options['hooks']->dispatch('requests.before_parse', [&$response, $url, $headers, $data, $type, $options]);
+			$parsed_response = self::parse_response($response, $url, $headers, $data, $options);
+		} catch (Exception $e) {
+			$options['hooks']->dispatch('requests.failed', [$e, $url, $headers, $data, $type, $options]);
+			throw $e;
+		} catch (InvalidArgument $e) {
+			$options['hooks']->dispatch('requests.failed', [$e, $url, $headers, $data, $type, $options]);
+			throw $e;
+		}
```

Comment:
Is there a need to catch these exceptions individually?

Reply:
No, simply evolved that way from writing and then fixing the tests. I combined it now